lib/Wrapper/OptimizingScheduler.cpp

```diff
@@ -237,6 +237,11 @@ void ScheduleDAGOptSched::SetupLLVMDag() {
 
   // Apply llvm DAG post processing.
   if (EnableMutations) {
+    addMutation(createCopyConstrainDAGMutation(TII, TRI));
```

Comment:
I think we should follow how LLVM add mutations to the DAG. https://github.com/llvm/llvm-project/blob/master/llvm/lib/Target/X86/X86TargetMachine.cpp#L350

Create the DAG, add mutations, then return the DAG object pointer. The AMDGPU target also uses `SetupLLVMDag()` and we don't want this mutation added to the AMDGPU target. Doing it the LLVM way would make the mutations target-dependent.

Reply:
Thanks for the reply.  I have checked the code, and you are right the AMD scheduler doesn't use the generic LLVM mutations.  I didn't want to add it to createOptSched, b/c it is called before the constructor, and consequently before the sched.ini options including LLVM_MUTATIONS are loaded.  The constructor is not a good place b/c it is used by the AMDGPU target as well.  initSchedulers() (the place where the AMDGPU target adds its mutations) is also not a place b/c init schedulers is only called in the two pass configuration.
We could also just remove the LLVM_MUTATIONS option and add the generic mutation to createOptSched() or we could add a virtual function that is responsible for adding LLVM mutations if the option is present.  I am open to any of the aforementioned options.

Reply:
I don't see a problem with adding the mutations before LLVM_MUTATIONS is loaded. They won't get applied until we call postprocessDAG() in SetupLLVMDag() anyways. By then LLVM_MUTATIONS should be loaded.

Either way I'm fine as long as the mutations doesn't get applied to all targets.

Reply:
ScheduleDAGMILive::schedule() calls postProcessDAG(). There are two steps to implementing the mutations. They are added via addMutation, but are not actually applied until postProcessDAG is called. For this reason I thought it best to avoid calling SetupLLVMDag if we have already called the generic scheduler. I'm not sure postProcessDAG is only called once in all permutations.

Reply:
Thanks for bringing that up! That is a good point that I have missed. We don't need the call to `SetupLLVMDag()` before the generic scheduler if the generic scheduler itself already calls `postProcessDAG()`.

Reply:
@vangthao95  I was going to check if ScheduleDAGMILive::schedule() if or any other part of the LLVM code base called postProcessDAG(), when @kerbowa posted his last message.